daliuge-engine/dlg/apps/app_base.py

```diff
@@ -612,7 +616,24 @@ def _run(self):
             #         attr_name, load_pickle(named_inputs[attr_name])
             #     )
             # else:
-            self.__setattr__(attr_name, named_inputs[attr_name])
+            if isinstance(named_inputs[attr_name], list) and len(named_inputs[attr_name]) > 1:
+                for ni in named_inputs[attr_name]:
+                    if ni.status != DROPStates.COMPLETED:
+                        continue
+                    break # we use the first completed
+            else:
+                ni = named_inputs[attr_name]
+            logger.debug("Identified input: %s", ni.name)
+            if  "componentParams" not in ni.parameters or ni.parameters["componentParams"]["dropclass"]["value"] != "dlg.data.drops.data_base.NullDROP" and ni.status == DROPStates.COMPLETED:
```

Comment:
I think we need to add further documentation to this method to explain what is going on here. I don't quite understand why we need to check the parsing here?